subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultFileSystemAccess.java

```diff
@@ -93,23 +93,25 @@ public <T> Optional<T> readRegularFileContentHash(String location, Function<Hash
                 }
                 return Optional.empty();
             })
-            .orElseGet(() -> {
+            .orElseGet(() -> virtualFileSystem.store(location, vfsStorer -> {
                 File file = new File(location);
                 FileMetadata fileMetadata = this.stat.stat(file);
                 if (fileMetadata.getType() == FileType.Missing) {
-                    virtualFileSystem.store(location, () -> new MissingFileSnapshot(location, fileMetadata.getAccessType()));
+                    vfsStorer.accept(new MissingFileSnapshot(location, fileMetadata.getAccessType()));
                 }
                 if (fileMetadata.getType() != FileType.RegularFile) {
                     return Optional.empty();
                 }
                 HashCode hash = producingSnapshots.guardByKey(location,
                     () -> virtualFileSystem.findSnapshot(location)
-                        .orElseGet(() -> virtualFileSystem.store(location, () -> {
+                        .orElseGet(() -> {
                             HashCode hashCode = hasher.hash(file, fileMetadata.getLength(), fileMetadata.getLastModified());
-                            return new RegularFileSnapshot(location, file.getName(), hashCode, fileMetadata);
-                        })).getHash());
+                            RegularFileSnapshot fileSnapshot = new RegularFileSnapshot(location, file.getName(), hashCode, fileMetadata);
+                            vfsStorer.accept(fileSnapshot);
+                            return fileSnapshot;
+                        })).getHash();
```

Comment:
We could return the hash itself from the lambda now, and I think that'd make the code simpler to understand.

Reply:
To me this looks simpler:

```java
    @Override
    public <T> Optional<T> readRegularFileContentHash(String location, Function<HashCode, T> visitor) {
        return virtualFileSystem.findMetadata(location)
            .<Optional<FileSystemLocationSnapshot>>flatMap(snapshot -> {
                if (snapshot.getType() != FileType.RegularFile) {
                    return Optional.of(Optional.empty());
                }
                if (snapshot instanceof FileSystemLocationSnapshot) {
                    return Optional.of(Optional.of((FileSystemLocationSnapshot) snapshot));
                }
                return Optional.empty();
            })
            .orElseGet(() -> virtualFileSystem.store(location, vfsStorer -> {
                File file = new File(location);
                FileMetadata fileMetadata = this.stat.stat(file);
                if (fileMetadata.getType() == FileType.Missing) {
                    MissingFileSnapshot snapshot = new MissingFileSnapshot(location, fileMetadata.getAccessType());
                    vfsStorer.store(snapshot);
                    return Optional.of(snapshot);
                }
                if (fileMetadata.getType() != FileType.RegularFile) {
                    return Optional.empty();
                }
                FileSystemLocationSnapshot snapshot = producingSnapshots.guardByKey(location,
                    () -> virtualFileSystem.findSnapshot(location)
                        .orElseGet(() -> {
                            HashCode hashCode = hasher.hash(file, fileMetadata.getLength(), fileMetadata.getLastModified());
                            return vfsStorer.store(new RegularFileSnapshot(location, file.getName(), hashCode, fileMetadata));
                        }));
                return Optional.of(snapshot);
            }))
            .map(FileSystemLocationSnapshot::getHash)
            .map(visitor);
    }
```